Passing a function as an argument: C error. This should never happen.

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V doctor:** 
```OS: linux, "Artix Linux"
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 3600 6-Core Processor
CC version: cc (GCC) 11.1.0

getwd: /home/me/coding/v/nice_server
vmodules: /home/me/.vmodules
vroot: /home/me/Programs/v
vexe: /home/me/Programs/v/v
vexe mtime: 2021-05-30 12:11:06
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 e4f6369.aa861d1

Git version: git version 2.31.1
Git vroot status: weekly.2021.21-60-gaa861d1c
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v run .`

**What did you expect to see?**
Compilation
 
**What did you see instead?**
```==================
/tmp/v/nice_server.13345884300252405638.tmp.c:1557: error: ';' expected (got "main__on_connection")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

**What caused this?**
Passing the function `on_connection` seems to have caused this.

**Source files:**
`example_server.v`
```
/////////////////////////////// example server

import nice_server

import os

const (
	inp_port = 8100
	inp_header_maxlen = 512 // bytes
)

fn main() {
	
	mut server := nice_server.new_server(inp_port, inp_header_maxlen)

	server.initialise() ?
	defer { // this will get run last
		server.deinitialise() or{
			println('could not deinitialise: $err')
		}
	}
	
	server_thread := go (&server).start(on_connection)
	defer { // this will get run second
		server_thread.wait()
	}

	println('press ENTER to stop the server')
	os.input('')
	defer { // this will get run first
		server.stop() or{
			println('could not stop: $err')
		}
	}

}


fn on_connection(c nice_server.Client){
	println('bruh')
}
```
`nice_server.v`
```
module nice_server


pub fn new_server(port int, header_maxlen int) Server {
	return Server{
		port: port
		header_maxlen: header_maxlen
	}
}


// errors
enum Err{
	no_err = 0

	// receiving
	cant_listen
	cant_accept
	cant_get_ip
	cant_get_addr
	cant_read

	// header
	didnt_receive_full_header
	header_is_len0
	bad_header_metadata

	// sending
	cant_send_response

	// quitting
	cant_close
	cant_close_listener
}

fn raise(info string, code Err) ? {
	return error_with_code(info, int(code))
}
```
`nice_server_class.v`
```
module nice_server

import net

struct Server{
	port int
	header_maxlen int
mut:
	listener net.TcpListener

	running bool
}

fn (mut s Server) raise(info string, code Err) ? {
	raise('server error: $info', code) ?
}


pub fn (mut s Server) initialise() ? {
	s.listener = net.listen_tcp(s.port) or{
		return s.raise('cannot listen on port $s.port', Err.cant_listen)
	}
	s.running = true
	println('running on port $s.port')
}

pub fn (mut s Server) deinitialise() ? {
	s.listener.close() or{
		return s.raise('cannot close the listener', Err.cant_close_listener)
	}
}


pub fn (mut s Server) start(response_fnc fn(Client)){
	for s.running {
		s.handle_next_connection(response_fnc) or{
			println('msg fucked up, idk: $err')
		}
	}
}

pub fn (mut s Server) stop() ? {
	s.running = false
}


// TODO add timeouts
fn (mut s Server) handle_next_connection(response_fnc fn(Client)) ? {

	// TODO add a timeout especially here
	mut con := s.listener.accept() or{
		return s.raise('cannot accept', Err.cant_accept)
	}

	ip := con.peer_ip() or{
		return s.raise('cannot get ip', Err.cant_get_ip)
	}
	addr := con.peer_addr() or{
		return s.raise('cannot get addr', Err.cant_get_addr)
	}

	mut client := Client{
		ip: ip
		addr: addr
		header_maxlen: s.header_maxlen
		con: con
		running: &s.running
	}

	go (&client).handle_connection(response_fnc)
}
```
`nice_client_class.v`
```
module nice_server

import net

struct Client{
	ip string
	addr net.Addr
	header_maxlen int	
	running &bool
mut:
	con net.TcpConn
}

fn (c Client) raise(info string, code Err) ? {
	return raise('An error occured in client [$c.ip ; $c.addr]: $info', code)
}

fn (mut c Client) handle_connection(response_fnc fn(Client)) ? {

	/* //this causes -> Segmentation fault
	defer{
		c.con.close() or{
			c.raise('couldnt close connection', Err.cant_close) ?
		}
	}
	*/

	println('accepted connection from $c.ip')
	println('with an addr $c.addr')

	mut data := []byte{len: c.header_maxlen}
	data_len := c.con.read(mut data) or{
		return c.raise('cannot read data', Err.cant_read)
	}
	data = data[..data_len]

	mut data_str := data.bytestr()

	if ! data_str.ends_with('\r\n\r\n') {
		c.raise('didnt reveive full header', Err.didnt_receive_full_header) ?
	}
	data_str = data_str[..data_str.len - 4]

	mut header := data_str.split('\r\n')
	if header.len == 0 {
		c.raise('header is of length 0', Err.header_is_len0) ?
	}

	metadata := header[0]
	if metadata.count(' ') != 2 {
		c.raise('bad header metadata', Err.bad_header_metadata) ?
	}
	header = header[1..]

	split_metadata := metadata.split(' ')
	request_type := split_metadata[0]
	requested_page := split_metadata[1] // deal with '/../..'
	http_version := split_metadata[2]

	//println('type: $request_type')
	//println('page: $requested_page')
	//println('http: $http_version')

	// TODO deal with the rest of the header -> make it into a map
	//println('header remains: $header')

	response_fnc(c)

	sent_data_str := 'You requested a $request_type version of $requested_page, at $http_version'
	sent_data := sent_data_str.bytes()
	sent_len := c.con.write(sent_data) or{
		return c.raise('could not send response', Err.cant_send_response)
	}
	assert sent_len == sent_data.len

	// TODO make this into a defer (after getting the seg fault fixed)
	c.con.close() or{
		c.raise('couldnt close connection', Err.cant_close) ?
	}
}
```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1826,8 +1826,14 @@ fn (mut g Gen) go_expr(node ast.GoExpr) {
 			g.type_definitions.writeln('EMPTY_STRUCT_DECLARATION;')
 		} else {
 			for i, arg in expr.args {
-				styp := g.typ(arg.typ)
-				g.type_definitions.writeln('\t$styp arg${i + 1};')
+				arg_sym := g.table.sym(arg.typ)
+				if arg_sym.info is ast.FnType {
+					sig := g.fn_var_signature(arg_sym.info.func.return_type, arg_sym.info.func.params, 'arg${i + 1}')
+					g.type_definitions.writeln('\t' + sig + ';')
+				} else {
+					styp := g.typ(arg.typ)
+					g.type_definitions.writeln('\t$styp arg${i + 1};')
+				}
 			}
 		}
 		if need_return_ptr {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1828,7 +1828,8 @@ fn (mut g Gen) go_expr(node ast.GoExpr) {
 			for i, arg in expr.args {
 				arg_sym := g.table.sym(arg.typ)
 				if arg_sym.info is ast.FnType {
-					sig := g.fn_var_signature(arg_sym.info.func.return_type, arg_sym.info.func.params, 'arg${i + 1}')
+					sig := g.fn_var_signature(arg_sym.info.func.return_type, arg_sym.info.func.params,
+						'arg${i + 1}')
 					g.type_definitions.writeln('\t' + sig + ';')
 				} else {
 					styp := g.typ(arg.typ)

--- a/vlib/v/tests/go_call_fn_with_anon_fn_arg_test.v
+++ b/vlib/v/tests/go_call_fn_with_anon_fn_arg_test.v
@@ -1,0 +1,14 @@
+fn start(f fn ()) string {
+	f()
+	return 'ok!!'
+}
+
+fn on_connect() {
+	println('fn ok!!')
+}
+
+fn test_go_call_fn_with_anon_fn_arg() {
+	g := go start(on_connect)
+	ret := g.wait()
+	assert ret == 'ok!!'
+}

